### PR TITLE
Export and import schema and data file together

### PR DESCRIPTION
**Pull request by flyingsilverfin**
## What is the goal of this PR?

We improve the UX of importing and exporting data from TypeDB: previously, we had to independently export the schema through Console, and then create the database and define the schema through Console at import time.

Now both the import and export commands take two flags: `--schema` and `--data`, which for export specify where to write the schema and data files to, and for import specify where to read the schema and data files from.

We implement validation to ensure that the database that is being imported into does not previously exist. Before, we allowed loading on top of an existing database, which could lead to unpredictable behaviour.

Closes https://github.com/vaticle/typedb/issues/6774

## What are the changes implemented in this PR?

* Update the `typedb export` and `typedb import` commands to accept separate parameters for schema and data
* Add new validation to only allow importing into a database that does not exist previously
* Refactor internal architecture to use the term `Database` migration instead of `Data` migration, since it now includes both Schema and Data

### CODE REVIEW

File: migrator/migrator.proto [hunk 1]
```diff
@@ -27,14 +27,15 @@ option java_outer_classname = "MigratorProto";
...
-    rpc ImportData(Import.Req) returns (stream Import.Progress);
+    rpc ExportDatabase(Export.Req) returns (stream Export.Progress);
+    rpc ImportDatabase(Import.Req) returns (stream Import.Progress);
 }
 
 message Export {
     message Req {
         string database = 1;
-        string filename = 2;
+        string data_file = 2;
```
Comment (flyingsilverfin) [inline migrator/migrator.proto:37]: useful to maintain backwards compat so users can still use the same migrator to import into the new database on a different TypeDB server

File: migrator/CoreMigratorClient.java [hunk 2]
```diff
@@ -46,24 +46,26 @@ public static CoreMigratorClient create(int serverPort) {
         return new CoreMigratorClient(stub);
     }
 
-    public boolean importData(String database, Path file) {
+    public boolean importDatabase(String database, Path schemaFile, Path dataFile) {
```
Comment (flyingsilverfin) [inline migrator/CoreMigratorClient.java:49]: new consistent terminology

File: migrator/MigratorService.java [hunk 3]
```diff
@@ -58,10 +61,13 @@ public void exportData(MigratorProto.Export.Req request, StreamObserver<Migrator
...
 
     @Override
-    public void importData(MigratorProto.Import.Req request, StreamObserver<MigratorProto.Import.Progress> responseObserver) {
-        DataImporter importer = null;
+    public void importDatabase(MigratorProto.Import.Req request, StreamObserver<MigratorProto.Import.Progress> responseObserver) {
+        DatabaseImporter importer = null;
         try {
-            importer = new DataImporter(databaseMgr, request.getDatabase(), Paths.get(request.getFilename()), version);
+            importer = new DatabaseImporter(
+                    databaseMgr, request.getDatabase(), Paths.get(request.getSchemaFile()),
```
Comment (flyingsilverfin) [inline migrator/MigratorService.java:68]: new parameter everywhere: `schemaFile`

File: migrator/database/DatabaseExporter.java [hunk 4]
```diff
@@ -44,32 +45,48 @@
...
+        this.dataFile = dataFile;
         this.version = version;
         this.status = new Status();
     }
 
     public void run() {
         LOG.info("Exporting {} from TypeDB {}", database, version);
-        try (OutputStream outputStream = new BufferedOutputStream(Files.newOutputStream(filename))) {
+        writeSchema();
+        writeData();
```
Comment (flyingsilverfin) [inline migrator/database/DatabaseExporter.java:76]: split two phases explicitly

File: migrator/database/DatabaseImporter.java [hunk 5]
```diff
@@ -75,26 +79,32 @@
...
+    private final String database;
+    private final Path schemaFile;
+    private TypeDB.Session session;
 
-    public DataImporter(TypeDB.DatabaseManager typedb, String database, Path dataFile, String version) {
+    public DatabaseImporter(TypeDB.DatabaseManager databaseMgr, String database, Path schemaFile, Path dataFile, String version) {
         if (!Files.exists(dataFile)) throw TypeDBException.of(FILE_NOT_FOUND, dataFile);
-        this.session = typedb.session(database, Arguments.Session.Type.DATA);
+        if (!Files.exists(schemaFile)) throw TypeDBException.of(FILE_NOT_FOUND, schemaFile);
+        if (databaseMgr.contains(database)) throw TypeDBException.of(DATABASE_EXISTS, database);
```
Comment (flyingsilverfin) [inline migrator/database/DatabaseImporter.java:104]: database pre-exists validation

File: migrator/migrator.proto (hunk 1 above)
Comment (benjaminasmall) [reply migrator/migrator.proto:37]: But it's a shame that this reverses the order of schema and data compared to everywhere else.

File: migrator/database/DatabaseImporter.java [hunk 6]
```diff
@@ -116,15 +128,28 @@ public void run() {
             if (!checksum.verify(status)) throw TypeDBException.of(IMPORT_CHECKSUM_MISMATCH, checksum.mismatch(status));
             Instant end = Instant.now();
             LOG.info("Finished in: " + Duration.between(start, end).getSeconds() + " seconds");
-            LOG.info("Imported: " + status.toString());
+            LOG.info("Imported: " + status);
```
Comment (benjaminasmall) [inline migrator/database/DatabaseImporter.java]: Would it be useful to give the database name, in parallel to giving the filenames on export?

File: server/parameters/CoreSubcommandParser.java [hunk 7]
```diff
@@ -121,13 +123,13 @@ public Import() {
...
-            return new CoreSubcommand.Import(database.parse(options).get(), filePath.parse(options).get(),
-                    port.parse(options).get());
+            return new CoreSubcommand.Import(database.parse(options).get(), schemaFile.parse(options).get(),
+                    dataFile.parse(options).get(), port.parse(options).get());
         }
 
         @Override
         public List<Help> helpList() {
-            return list(database.help(), filePath.help(), port.help());
+            return list(database.help(), dataFile.help(), port.help());
```
Comment (benjaminasmall) [inline server/parameters/CoreSubcommandParser.java:132]: `schamaFile.help()` too?

File: server/parameters/CoreSubcommandParser.java [hunk 8]
```diff
@@ -151,13 +155,13 @@ public Export() {
...
-            return new CoreSubcommand.Export(database.parse(options).get(), filePath.parse(options).get(),
-                    port.parse(options).get());
+            return new CoreSubcommand.Export(database.parse(options).get(), schemaFile.parse(options).get(),
+                    dataFile.parse(options).get(), port.parse(options).get());
         }
 
         @Override
         public List<Help> helpList() {
-            return list(database.help(), filePath.help(), port.help());
+            return list(database.help(), dataFile.help(), port.help());
```
Comment (benjaminasmall) [inline server/parameters/CoreSubcommandParser.java:164]: Same.

File: server/parameters/CoreSubcommandParser.java [hunk 9]
```diff
@@ -108,10 +108,12 @@ public static class Import extends SubcommandParser<CoreSubcommand.Import> {
 
         public static final OptionParser.String database =
                 new OptionParser.String("database", "Database to import into.");
-        public static final OptionParser.Path filePath =
-                new OptionParser.Path("file", "Path to data file to import (.typedb format).");
+        public static final OptionParser.Path schemaFile =
+                new OptionParser.Path("schema", "Path to schema file to import.");
```
Comment (benjaminasmall) [inline server/parameters/CoreSubcommandParser.java:112]: Specify format as `.tql`, as below?

**Review by benjaminasmall — CHANGES_REQUESTED** [review]

File: migrator/database/DatabaseImporter.java (hunk 6 above)
Comment (flyingsilverfin) [reply migrator/database/DatabaseImporter.java]: Can add

File: server/parameters/CoreSubcommandParser.java (hunk 7 above)
Comment (flyingsilverfin) [reply server/parameters/CoreSubcommandParser.java:132]: good catch!

File: server/parameters/CoreSubcommandParser.java (hunk 9 above)
Comment (flyingsilverfin) [reply server/parameters/CoreSubcommandParser.java:112]: sure

**Review by benjaminasmall — APPROVED** [review]

**Pull request merged**